File: src/main/java/codingdojo/Store.java
package codingdojo;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Represents a physical Store where you can go and buy
 * products and attend events.
 */
public class Store implements ModelObject {

    private final Map<String, Item> itemsInStock = new HashMap<>();
    private final String name;
    private boolean droneDelivery;

    public Store(String name, boolean droneDelivery) {
        this.name = name;
        this.droneDelivery = droneDelivery;
    }
    public void addStockedItems(Item... items) {
        for (Item item: items) {
            this.itemsInStock.put(item.getName(), item);
        }
    }

    public void addStoreEvent(StoreEvent storeEvent) {
        this.itemsInStock.put(storeEvent.getName(), storeEvent);
    }

    public void removeStockedItems(Item... items) {
        for (Item item: items) {
            this.itemsInStock.remove(item.getName());
        }
    }

    public boolean hasItem(Item item) {
        return itemsInStock.containsKey(item.getName());
    }

    public Item getItem(String name) {
        return itemsInStock.get(name);
    }

    public boolean hasDroneDelivery() {
        return droneDelivery;
    }

    @Override
    public String toString() {
        return "Store{" +
                "name='" + name + "\', " +
                "droneDelivery=" + droneDelivery +
                '}';
    }

    @Override
    public void saveToDatabase() {
        throw new UnsupportedOperationException("missing from this exercise - shouldn't be called from a unit test");
    }

}

File: typescript/src/Store.ts
import {ModelObject} from './ModelObject'
import Item from './Item';
import StoreEvent from './StoreEvent';
import { UnsupportedOperationException } from './Errors';

/**
 * Represents a physical Store where you can go and buy
 * products and attend events.
 */
export default class Store implements ModelObject {
    private itemsInStock: Map<string, Item> = new Map();
    private readonly name: string;
    private droneDelivery: boolean;

    constructor(name: string, droneDelivery: boolean) {
        this.name = name;
        this.droneDelivery = droneDelivery;
    }
    addStockedItems(...items: Item[]) {
        for (let item of items) {
          this.itemsInStock.set(item.getName(), item);
        }
    }

    addStoreEvent(storeEvent: StoreEvent) {
        this.itemsInStock.set(storeEvent.getName(), storeEvent);
    }

    removeStockedItems(...items: Item[]) {
        for (let item of items) {
            this.itemsInStock.delete(item.getName());
        }
    }

    hasItem(item: Item) {
        return this.itemsInStock.has(item.getName());
    }

    getItem(name: string) {
        return this.itemsInStock.get(name);
    }

    hasDroneDelivery() {
        return this.droneDelivery;
    }

    toString() {
        return "Store{" +
                "name='" + this.name + "\', " +
                "droneDelivery=" + this.droneDelivery +
                '}';
    }

    saveToDatabase() {
        throw new UnsupportedOperationException("missing from this exercise - shouldn't be called from a unit test");
    }

}
